### ISSUE TITLE
When will the Quantization of YoloV8 come out？

### ISSUE THREAD

**Issue by sdreamforchen** [ultralytics/ultralytics#1650]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When will the Quantization of YoloV8 come out？

### Additional

Thanks a lot！

**Comment by glenn-jocher** [comment]
@sdreamforchen quantization is already supported by many export formats, i.e.

```bash
# FP16 quantization
yolo export format=coreml half

# int8 quantization
yolo export format=coreml int8
```

**Comment by sdreamforchen** [comment]
Can you share the mAP of n/s/m model based on int8



---Original---
From: "Glenn ***@***.***&gt;
Date: Mon, Mar 27, 2023 20:57 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] When will the Quantization of YoloV8 come out？ (Issue #1650)




 
@sdreamforchen quantization is already supported by many export formats, i.e.
 # FP16 quantization yolo export format=coreml half # int8 quantization yolo export format=coreml int8

**Comment by glenn-jocher** [comment]
@sdreamforchen you can val any exported model yourself to get any metrics you require. See https://docs.ultralytics.com/modes/val/

**Comment by sdreamforchen** [comment]
> @sdreamforchen you can val any exported model yourself to get any metrics you require. See https://docs.ultralytics.com/modes/val/

The mAP is very close to Float32's mAP, what's your Quantization strategy? Full or partial

**Comment by glenn-jocher** [comment]
@sdreamforchen YOLOv8 uses dynamic quantization (partially quantized) with PyTorch.

**Comment by victorsoyvictor** [comment]
Curious thing happening when exporting 3 variants of onnx model:
```
yolo export model=best.pt opset=12 format=onnx imgsz=640,640 int8
yolo export model=best.pt opset=12 format=onnx imgsz=640,640 half
yolo export model=best.pt opset=12 format=onnx imgsz=640,640 
```
I get binary the same (compare HEX) between "normal" and int8. Half is halved though and I am happy with it.

Thanks,

**Comment by glenn-jocher** [comment]
@victorsoyvictor it's normal to get the same binary between the "normal" (float32) and int8 when exporting an ONNX model. In dynamic quantization, the activation densities are tracked for every layer during training, so when inferring, the model converts float32 input tensors to float16 depthwise convolution weights and int8 quantized pointwise weights on-the-fly. This reduces memory bandwidth, allowing faster inference than pure float32 computation, while still giving similar model accuracies.

**Comment by victorsoyvictor** [comment]
I'm using Yolov8CPPInference (uses OpenCV::DNN) example for getting an idea of the inference latency of a pretrained model (x6) trained on a custom dataset (which I'm happy in accuracy).

I have the fp16 fp32 and int8 ONNX versions obtained from Export tool (is a 640,640) and it seems to be a 80ms latency with FP16 and 190ms in FP32. 

How do I set INT8 in this CPP example? (For FP16 I use setpreferableTarget DNN CUDA FP16) I would like to see something fast. 

I'm running on a GTX3060 with CUDA. Could also check in GTX3090.

Thanks,

**Comment by HissaanAli** [comment]
It's 

> @sdreamforchen quantization is already supported by many export formats, i.e.
> 
> ```shell
> # FP16 quantization
> yolo export format=coreml half
> 
> # int8 quantization
> yolo export format=coreml int8
> ```

It's worth noting that this would only work on macOs. It won't work on Ubuntu or windows.

**Comment by khanonenet** [comment]
> > @sdreamforchen quantization is already supported by many export formats, i.e.
> > ```shell
> > # FP16 quantization
> > yolo export format=coreml half
> > 
> > # int8 quantization
> > yolo export format=coreml int8
> > ```
> 
> It's worth noting that this would only work on macOs. It won't work on Ubuntu or windows.

so there is no support on ubuntu/windows?

**Comment by glenn-jocher** [comment]
@khanonenet quantization support for YOLOv8 is not limited to macOS. The commands provided in the example are specific to exporting to the CoreML format for macOS, but quantization is supported in other export formats as well. For Ubuntu and Windows users, you can export the YOLOv8 model using different formats such as ONNX or TensorFlow, and then apply quantization techniques specific to those frameworks.

For example, in TensorFlow, you can use TensorFlow Lite's post-training quantization techniques to quantize the exported model. Similarly, in ONNX, you can apply quantization techniques using suitable libraries or frameworks.

While the specific commands mentioned in the example may not work on Ubuntu or Windows, there are alternative ways to achieve quantization on these platforms. You might need to explore the specific tools and frameworks available for your chosen export format.

Please let me know if you have any further questions or need additional assistance.

**Comment by absmahi01** [comment]
@glenn-jocher can you please give an example of tensorflow lite post training quantization?

**Comment by victorsoyvictor** [comment]
> I'm using Yolov8CPPInference (uses OpenCV::DNN) example for getting an idea of the inference latency of a pretrained model (x6) trained on a custom dataset (which I'm happy in accuracy).
> 
> I have the fp16 fp32 and int8 ONNX versions obtained from Export tool (is a 640,640) and it seems to be a 80ms latency with FP16 and 190ms in FP32.
> 
> How do I set INT8 in this CPP example? (For FP16 I use setpreferableTarget DNN CUDA FP16) I would like to see something fast.
> 
> I'm running on a GTX3060 with CUDA. Could also check in GTX3090.
> 
> Thanks,

Any help with CPP INT8 ? I would like to see somewhat fast inference.

**Comment by glenn-jocher** [comment]
@victorsoyvictor, for INT8 inference with OpenCV's DNN module and YOLOv8 ONNX models, OpenCV does not provide direct support for INT8 inference on GPU backends. The int8 quantized models are typically optimized to run on CPUs or specialized hardware that supports int8 operations. The CUDA backend for OpenCV DNN currently supports FP32 and FP16 inference. For fast inference on a GTX 3060 or GTX 3090, sticking to FP16 with `setPreferableTarget(cv::dnn::DNN_TARGET_CUDA_FP16)` is your best bet. 

For INT8 inference on GPU, consider using a framework that natively supports INT8 on CUDA like TensorRT. You would need to convert the ONNX model to a TensorRT optimized model to leverage INT8 precision on GPUs.
